src/aspire/abinitio/commonline_sync3n.py

```diff
@@ -0,0 +1,1017 @@
+import logging
+import os.path
+
+import numpy as np
+from numpy.linalg import norm
+from scipy.optimize import curve_fit
+
+from aspire.abinitio import CLOrient3D, SyncVotingMixin
+from aspire.utils import J_conjugate, all_pairs, nearest_rotations, tqdm, trange
+from aspire.utils.matlab_compat import stable_eigsh
+from aspire.utils.random import randn
+
+logger = logging.getLogger(__name__)
+
+
+class CLSync3N(CLOrient3D, SyncVotingMixin):
+    """
+    Define a class to estimate 3D orientations using common lines Sync3N methods (2017).
+
+    Ido Greenberg, Yoel Shkolnisky,
+    Common lines modeling for reference free Ab-initio reconstruction in cryo-EM,
+    Journal of Structural Biology,
+    Volume 200, Issue 2,
+    2017,
+    Pages 106-117,
+    ISSN 1047-8477,
+    https://doi.org/10.1016/j.jsb.2017.09.007.
+    """
+
+    # Initialize alternatives
+    #
+    # When we find the best J-configuration, we also compare it to the alternative 2nd best one.
+    # this comparison is done for every pair in the triplet independently. to make sure that the
+    # alternative is indeed different in relation to the pair, we document the differences between
+    # the configurations in advance:
+    # ALTS(:,best_conf,pair) = the two configurations in which J-sync differs from best_conf in relation to pair
+
+    _ALTS = np.array(
+        [
+            [[1, 2, 1], [0, 2, 0], [0, 0, 1], [1, 0, 0]],
+            [[2, 3, 3], [3, 3, 2], [3, 1, 3], [2, 1, 2]],
+        ],
+        dtype=int,
+    )
+
+    def __init__(
+        self,
+        src,
+        n_rad=None,
+        n_theta=360,
+        max_shift=0.15,
+        shift_step=1,
+        epsilon=1e-2,
+        max_iters=1000,
+        seed=None,
+        mask=True,
+        S_weighting=False,
+        J_weighting=False,
+        hist_intervals=100,
+    ):
+        """
+        Initialize object for estimating 3D orientations.
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction
+        :param n_theta: The number of points in the theta direction
+        :param max_shift: Maximum range for shifts as a proportion of resolution. Default = 0.15.
```

Comment:
Yeah I just copied these from our existing CL code. We almost have the complete set of CL codes at which point it might make sense to refactor them.

Changed:

```
        :param max_shift: Maximum range for shifts as a proportion of box size. Default = 0.15.                 
        :param shift_step: Step size of shift estimation in pixels. Default = 1 pixel.
```

Reply:
Thanks!